private-identity-audit/src/lib.rs

```diff
@@ -25,22 +37,107 @@ use uuid::Uuid;
 /// element of a set, PUIS can guess that element with probability 1/the_size_of_the_padded_set.
 pub const SET_SIZE_ANONYMITY_PARAM: usize = 100_000;
 
+pub struct InitialProver;
+pub struct FinalProver;
+
+pub struct VerifierSetGenerator;
+pub struct Verifier;
+
 /// The initial private set of PUIS.
-pub type PrivateUids = Vec<Scalar>;
+pub struct PrivateUids(pub Vec<Scalar>);
+
+impl Encode for PrivateUids {
+    #[inline]
+    fn size_hint(&self) -> usize {
+        32 * self.0.len()
+    }
+
+    fn encode_to<W: Output>(&self, dest: &mut W) {
+        let uuids: Vec<[u8; 32]> = self
+            .0
+            .clone()
+            .into_iter()
+            .map(|u| *u.as_bytes())
+            .collect::<Vec<_>>();
+
+        uuids.encode_to(dest);
+    }
+}
+
+impl Decode for PrivateUids {
+    fn decode<I: Input>(input: &mut I) -> Result<Self, CodecError> {
+        let uuids = <Vec<[u8; 32]>>::decode(input)?;
+
+        let uuids: Vec<Scalar> = uuids.into_iter().map(Scalar::from_bits).collect::<Vec<_>>();
+
+        Ok(PrivateUids(uuids))
+    }
+}
 
 /// The committed and padded version of the private set of PUIS.
-pub type CommittedUids = Vec<RistrettoPoint>;
+#[derive(Clone, Debug, PartialEq)]
+pub struct CommittedUids(pub Vec<RistrettoPoint>);
+
+impl Encode for CommittedUids {
+    #[inline]
+    fn size_hint(&self) -> usize {
+        32 * self.0.len()
+    }
+
+    fn encode_to<W: Output>(&self, dest: &mut W) {
+        let committed_uuids: Vec<[u8; 32]> = self
+            .0
+            .clone()
+            .into_iter()
+            .map(|u| *u.compress().as_bytes())
+            .collect::<Vec<_>>();
```

Comment:
Performance: Remove some intermediate copies.
```suggestion
        let committed_uuids = self.0.iter().map(|u| u.compress().to_bytes()).collect::<Vec<_>>();
```